concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs

```diff
@@ -0,0 +1,737 @@
+
+-- |Purge transactions from the transaction table and pending transactions.
+-- Transactions are purged only if they are not included in a live block, and
+-- have either expired or arrived longer ago than the transaction keep alive time.
+--
+-- If the first argument is @False@, the transaction table is only purged if
+-- 'rpInsertionsBeforeTransactionPurged' transactions have been inserted since
+-- the last purge.  If it is true, the table is purged regardless.
+--
+--   * Every 'BlockItem' in the transaction table that is not included in a live
+--     or finalized block is referenced in the pending transaction table.  That is,
+--     for a basic transaction the '_pttWithSender' table contains an entry for
+--     the sender where the nonce of the transaction falls within the range,
+--     and for a credential deployment the transaction hash is included in '_pttDeployCredential'.
+--
+--   * The low nonce for each entry in '_pttWithSender' is at least the last finalized
+--     nonce recorded in the account's non-finalized transactions in the transaction
+--     table.
+--
+--   * The pending transaction table only references transactions that are in the
+--     transaction table.  That is, the high nonce in a range is a tight bound and
+--     the deploy credential hashes correspond to transactions in the table.
+--
+--   * No non-finalized block is considered live or will become live if its round
+--     is less than or equal to the slot number of the last finalized block.
+--
+--   * If a transaction is known to be in any block that is not finalized or dead,
+--     then 'commitTransaction' or 'addCommitTransaction' has been called with a
+--     slot number at least as high as the slot number of the block.
+purgeTransactionTable ::
+    (MonadState (SkovData pv) m) =>
+    -- |Whether to force the purging.
+    Bool ->
+    -- |The current time.
+    UTCTime ->
+    m ()
+purgeTransactionTable force currentTime = do
+    purgeCount <- use transactionTablePurgeCounter
+    RuntimeParameters{..} <- use runtimeParameters
+    when (force || purgeCount > rpInsertionsBeforeTransactionPurge) $ do
+        transactionTablePurgeCounter .= 0
+        lfb <- use lastFinalized
+        let lastFinalizedRound = blockRound $! bpBlock lfb
+        transactionTable' <- use transactionTable
+        pendingTransactions' <- use pendingTransactionTable
+        let
+            currentTransactionTime = utcTimeToTransactionTime currentTime
+            oldestArrivalTime =
+                if currentTransactionTime > rpTransactionsKeepAliveTime
+                    then currentTransactionTime - rpTransactionsKeepAliveTime
+                    else 0
+            currentTimestamp = utcTimeToTimestamp currentTime
+            (newTT, newPT) = Purge.purgeTables (commitPoint lastFinalizedRound) oldestArrivalTime currentTimestamp transactionTable' pendingTransactions'
+        transactionTable .=! newTT
+        pendingTransactionTable .=! newPT
```

Comment:
It's not ideal no. I've created this issue for it and will get to it asap. 

https://github.com/Concordium/concordium-node/issues/768

The reason I'm not spending time on it in this PR is that I'm afraid it would make it drag on longer and currently there's a few PRs already: https://github.com/Concordium/concordium-node/pull/748, https://github.com/Concordium/concordium-node/pull/749,  https://github.com/Concordium/concordium-node/pull/751 "and more in progress"..  Which could benefit from this one being merged.